handling stopped packages for log and resource monitoring

jlurgo opened the pull request:
- watching /status/package topic to start/stop resource and logs monitoring
- adds initial update for status watch which was not happening based on changes in some cases

chfritz commented on robot-agent/logMonitor.js:
```diff
@@ -81,12 +81,51 @@ class LogMonitor {
+      
+      // Subscribe to package topics to automatically start/stop monitoring
+      this.mqttSync.data.subscribePathFlat(`${this.AGENT_PREFIX}/status/package`, (value, topic, matched) => {
+        const topicParts = topic.split('/');
+        // Extract package name from topic like "/agent-xxx/status/package/@scope/name/..."
+        const agentPrefixParts = this.AGENT_PREFIX.split('/').length;
+        if (topicParts.length < agentPrefixParts + 4) return; // Not enough parts for a valid package topic
+        
+        const packageScope = topicParts[agentPrefixParts + 2];
+        const packageName = topicParts[agentPrefixParts + 3];
+        const fullPackageName = `${packageScope}/${packageName}`;
```
It's better to use `parseTopic` here, rather than parsing the topic manually.

chfritz commented on robot-agent/logMonitor.js:
```diff
@@ -81,12 +81,51 @@ class LogMonitor {
-          this.watchLogs(packageName); // Start watching logs for the package
+      
+      // Subscribe to package topics to automatically start/stop monitoring
+      this.mqttSync.data.subscribePathFlat(`${this.AGENT_PREFIX}/status/package`, (value, topic, matched) => {
```
It this really the right topic to watch? `runningPackages` may be saver, because it comes directly from the local mqtt brokers on-client-connect logic.

chfritz replied:
The `startPackage` function (in main) handles this by first checking whether the package might already be running (by trying to find the PID). It only starts the package if it's not already running, i.e., when no such process can be found.

chfritz commented on robot-agent/logMonitor.js:
```diff
@@ -59,7 +59,7 @@ class LogMonitor {
     this.mqttClient = mqttClient;
     this.mqttSync = mqttSync;
     this.AGENT_PREFIX = agentPrefix;
-    this.mqttSync.subscribe(`${this.AGENT_PREFIX}/info`);
```
Why don't we need this anymore? Below we seem to be relying on it.

jlurgo replied:
since we're already publishing (in mqtt.js) on info topic, I though it was not needed to subscribe to it. I tested it and still works

chfritz commented on robot-agent/logMonitor.js:
```diff
@@ -81,12 +81,51 @@ class LogMonitor {
+      
+      // Check for existing package statuses at startup
+      const packageData = this.mqttSync.data.getByTopic(`${this.AGENT_PREFIX}/status/package`);
+      if (packageData) {
+        Object.keys(packageData).forEach(packageScope => {
+          if (typeof packageData[packageScope] === 'object') {
+            Object.keys(packageData[packageScope]).forEach(packageName => {
+              const packageStatus = packageData[packageScope][packageName];
+              if (packageStatus && typeof packageStatus === 'object' && packageStatus.status) {
+                const fullPackageName = `${packageScope}/${packageName}`;
+                log.info('Found existing running package, beginning log monitoring:', fullPackageName);
+                this.watchLogs(fullPackageName);
+              }
+            });
+          }
+        });
```
was this written by co-pilot by any chance? Please have a look at the existing functions in DataCache to traverse the data, e.g., https://transitiverobotics.com/docs/sdk/common/#formatch. My hunch is that one of those already implements the kind of traversal you need here.

chfritz commented on robot-agent/resourceMonitor.js:
```diff
@@ -116,6 +162,37 @@ class ResourceMonitor {
+      return;
+    }
+
+    // Use pgrep to find the PID of the running package
+    const { spawn } = require('child_process');
+    const pgrep = spawn('pgrep', ['-nf', `startPackage.sh ${packageName}`, '-U', process.getuid()]);
+    
+    let packagePid = null;
+    
+    pgrep.stdout.on('data', (data) => {
+      packagePid = parseInt(data.toString().trim());
+      log.debug(`Found running package ${packageName} with PID: ${packagePid}`);
+    });
+
+    pgrep.on('exit', (code) => {
+      if (code === 0 && packagePid) {
+        // Package is running, start monitoring
+        this.startMonitoring(packageName, packagePid);
+      } else {
+        log.debug(`Package ${packageName} not found running (exit code: ${code})`);
+      }
+    });
+
+    pgrep.on('error', (err) => {
+      log.error(`Error searching for package ${packageName}:`, err);
+    });
+  }
```
This seems to replicate logic we already have in mqtt.js for reliably getting the PID if the capability is running.

chfritz reviewed (commented):
Not sure I fully understand the need for these changes -- which are relatively big, and as you know agent changes carry a lot of danger, so we should be sure we need them. Can you elaborate on the problem this is solving? Is this to avoid the error we've seen re. no-PID for resource monitoring when the capability is stopped? If so then I think there is a much smaller solution where we just check runningPackages in the resource monitor before trying to look up the PID (or similar).

chfritz commented:
> The problem I'm solving here is that when you stop and restart a capability resource monitoring is lost, in addition to annoying error logging.

I think it's a good idea for the agent to always know which capabilities are running and what their respective PIDs are. When it starts them, it's easy because we can keep it in memory. But after a restart that is lost. I think it would be good to fix that, meaning to have the agent to keep in one place a list of all running capabilities incl. their PIDs no matter how and when they were started.

I see three options for that: either, on start, we just find all running packages once and add their status to an object where we keep track of them (which is also updated by the start/stop functions). Or, in the logic we use to track runningPackages, which has been working *very* reliably, we add logic to also look up the PID and then add that to the object. A third option would be to write the PID to mqttsync whenever we start a package. That way we can read that again after an agent restart.

Doing this centrally would allow us to avoid repeating some such logic in both the logMonitor and the resourceMonitor. We should also definitely avoid long reactive chains such as the one in this draft right now: package starts > package status is written to disk > a watchFile triggers > write that status to mqttSync > logMonitor and resourceMonitor react. Long chains like this introduce too many failure points, making them brittle -- for instance until recently the file-watching broke after an agent restart.

chfritz commented on robot-agent/packageMonitor.js:
```diff
@@ -0,0 +1,99 @@
+const log = getLogger('packageMonitor.js');
+log.setLevel('info');
+
+class PackageMonitor extends EventEmitter {
```
The big thing missing from this class is any info from `runningPackages` (or it's original source, namely clients connecting to the local mqtt broker). That is by far the most reliable source of truth about whether the package is running or not. So I find it confusing to introduce things like `isRunning` in this class that take a different meaning (namely whether we can find the PID).

chfritz commented on robot-agent/packageMonitor.js:
```diff
@@ -0,0 +1,99 @@
+  async initialize() {
+    await this.updateAllInstalledPackages();
+  }
+  async updateAllInstalledPackages() {
```
I found that function name confusing because "update" in the agent usually means updating the installed package itself (on disk), not the information about it. Maybe `refreshAllInstalledPackages`?

chfritz commented on robot-agent/packageMonitor.js:
```diff
@@ -0,0 +1,99 @@
+    await Promise.all(promises);
+    log.debug('Installed packages updated:', this.installedPackages);
+  }
+  async refreshSinglePackage(pkgName, currentInstalledPackages = null) {
```
Given our discussion about using `runningPackages` directly, I don't know that we even need this function.

chfritz reviewed (commented):
Some thoughts but as discussed today, I think we may want to rethink the approach and start with the `runningPackages` information from mqtt. This should also avoid the need to check which packages are running, because mqtt (from the LocalMQTT logic) should already tell us.

jlurgo commented:
I implemented what we discussed yesterday @chfritz , this is ready for another review pass

chfritz commented on robot-agent/packageMonitor.js:
```diff
@@ -0,0 +1,141 @@
+      async (value, topic, matched, tags) => {
+        const {scope, capName, version} = matched;
+        const pkgName = `${scope}/${capName}`;
+        await this.refreshAllInstalledPackages(); // ensure the package is in the list
```
Why refresh all and not just the one that just changed status?

chfritz commented on robot-agent/packageMonitor.js:
```diff
@@ -0,0 +1,141 @@
+          log.info(`Package ${pkgName} is running (via MQTT)`);
+          if (!this.installedPackages?.[pkgName]?.running) {
+            const pkgPid = await this.getPackagePid(pkgName);
+            this.installedPackages[pkgName] = this.installedPackages[pkgName] || {};
```
`this.installedPackages[pkgName] ||= {};`

chfritz commented on robot-agent/utils.js:
```diff
@@ -115,74 +113,36 @@ const killPackage = (name, signal = 'SIGTERM', cb = undefined) => {
-        });
-      subprocess.unref();
-      packagePid = subprocess.pid;
+  packageMonitor.waitForInitialization().then(() => {
```
why do we need to wait for initialization before starting packages? Packages are also started on agent-start and we'd like to avoid unnecessary delays there (such as waiting on a heartbeat from upstream).

jlurgo replied:
isn't it possible that, during startup, we call to start a package for which we still didn't receive the retained runningPackages message on mqtt? In that case we will start it twice

chfritz commented on robot-agent/packageMonitor.js:
```diff
@@ -0,0 +1,141 @@
+const log = getLogger('packageMonitor.js');
+log.setLevel('info');
+
+class PackageMonitor extends EventEmitter {
```
Please add doc-comments to this class and it's public methods.

chfritz reviewed (commented):
I think the new approach is much better! Some comments inline and I also get this when I start the agent (the named capability was already running when the agent started):

```
[2025-08-20T17:15:25.164Z utils debug] startPackage @transitive-robotics/configuration-management
**** Caught exception: TypeError: Cannot read properties of null (reading 'waitForHeartbeatOnce'): TypeError: Cannot read properties of null (reading 'waitForHeartbeatOnce')
    at /home/cfritz/work/transitive/transitive/robot-agent/packageMonitor.js:84:21
    at new Promise (<anonymous>)
    at PackageMonitor.waitForInitialization (/home/cfritz/work/transitive/transitive/robot-agent/packageMonitor.js:79:12)
    at startPackage (/home/cfritz/work/transitive/transitive/robot-agent/utils.js:121:18)
    at /home/cfritz/work/transitive/transitive/robot-agent/index.js:124:9
    at ChildProcess.exithandler (node:child_process:414:7)
    at ChildProcess.emit (node:events:524:28)
    at maybeClose (node:internal/child_process:1104:16)
    at ChildProcess._handle.onexit (node:internal/child_process:304:5)
```

jlurgo commented:
I just fixed the waiting for initialization error you found, thanks, more robust now
About that, I changed the method we had using pgrep on startPackage with the call to packageManager.
- advantages: 
  - uses runningPackages mqtt signal to tell what's running or not, which is always correct, using pgrep we have problems in dev (correct me if I'm wrong)
- disadvantage:
  - we have to wait until mqttsync is initialized and all runningPackages messages have arrived
  - one doubt here: when we use waitForHeartbeatOnce() is that expecting a hb from the cloud? if that's true this will only work with devices online... also true for logs and resources monitors whic wait for the same signal 

That said I can go back in utils.startPackage to using pgrep if you think it'd be better, we can keep the manager for starting/stopping logs and resource monitoring

chfritz commented on robot-agent/packageMonitor.js:
```diff
@@ -0,0 +1,168 @@
+  async refreshAllInstalledPackages() {
+    log.debug('Refreshing installed packages list...');
+    const installedPackages = getInstalledPackages();
+    installedPackages.map(async (pkgName) => {      
```
why `async `?

jlurgo commented on robot-agent/packageMonitor.js:
```diff
@@ -0,0 +1,168 @@
+   * 
+   * @returns {Promise<void>} Resolves when the monitor is initialized
+   */
+  waitForInitialization() {
```
yes it is, I left it because I liked the code and might be useful in the future... I can remove it

jlurgo commented on robot-agent/packageMonitor.js:
```diff
@@ -0,0 +1,168 @@
+   * Adds newly installed packages to the tracking list and removes uninstalled packages.
+   * If a package was running when it got uninstalled, emits a 'packageStopped' event.
+   */
+  async refreshAllInstalledPackages() {
```
fixed thanks!

chfritz commented on robot-agent/packageMonitor.js:
```diff
@@ -0,0 +1,168 @@
+    this.on('newListener', (event, listener) => {
+      log.debug(`New listener for event: ${event}`);
+      if (event === 'packageStarted') {
+        Object.entries(this.installedPackages).forEach(([pkgName, pkgData]) => {
```
We usually use `lodash` for this:
`_.forEach( this.installedPackages, (pkgData, pkgName) => ...`

Also, this and the `packageStopped` case are virtually the same.

chfritz commented on robot-agent/packageMonitor.js:
```diff
@@ -0,0 +1,168 @@
+      if (event === 'packageStarted') {
+        Object.entries(this.installedPackages).forEach(([pkgName, pkgData]) => {
+          if (pkgData.running) {
+            log.info(`Raising packageStarted event for already running package ${pkgName}`);
```
let's make this `debug` or just remove it -- seems a bit verbose

chfritz commented on robot-agent/packageMonitor.js:
```diff
@@ -0,0 +1,168 @@
+      }
+    );
+    this.refreshAllInstalledPackages();
+    this.mqttSync.waitForHeartbeatOnce(() => {
+      log.info('PackageMonitor heartbeat received');
+      this.initialized = true; // Set initialized state
+      log.info('PackageMonitor initialized');
+      this.emit('initialized');
+    });
```
This whole `initialized` logic is now obsolete it seems (no longer used anywhere if I see correctly).

chfritz commented on robot-agent/packageMonitor.js:
```diff
@@ -0,0 +1,168 @@
+        delete this.installedPackages[pkgName];
+        if (wasRunning) {
+          log.info(`Package ${pkgName} was detected uninstalled while running`);
+          this.emit('packageStopped', pkgName);
```
This doesn't seem right. Why would we emit this event here again? It's already being emitted (correctly) when runningPackages changes.

jlurgo replied:
that was for the case we find that an installed package that was running got removed...
Already removed all that

chfritz reviewed (commented):
More comments. But new high-level question: do we even need installedPackages in the PackageMonitor? It doesn't seem to serve any purpose anymore and I also don't know why would need it in first place. All we want to do is make sure that we start (resp. stop) the log- and resource monitors when a package starts (resp. stops), right?

chfritz commented on robot-agent/packageMonitor.js:
```diff
@@ -0,0 +1,105 @@
+    this.AGENT_PREFIX = agentPrefix;
+    log.info('Initializing PackageMonitor');
+    this.mqttSync.subscribe(`${this.AGENT_PREFIX}/status/runningPackages`);
+    this.mqttSync.data.subscribePathFlat(`${this.AGENT_PREFIX}/status/runningPackages/+scope/+capName/+version`,
```
before registering this subscription, don't we also need to check the current value of `runningPackages` and emit events accordingly?

But taking a step back for a second, I think this class has gone through a lot of changes and now, in the end, all we want it to do is:

- provide us with notifications of package start and stop -- which we can already get from `runningPackages`
- fetch the PID for packages when they start and send them along in the notification -- but for that we don't actually need to store it.

There is no need to introduce the notion of a late subscriber in this class, because it only relates to `runningPackages` in the mqttSync data (where the notion does make sense). But if this package just relays that info to its event listeners, then we'll never have "new" late listeners for which we would need to have stored the PID. So I think we may want to rethink this entire approach and what function, if any, this class should play.

As I mentioned, it is the agent itself who is writing `runningPackages`, so I think some of the issues that lead us on this roundabout course is that we didn't yet try the alternative I mentioned, namely to write into mqttSync directly in the localMqtt `clientReady` and `clientDisconnect` events. That would avoid any need for waiting for heartbeats from upstream and would also work offline.

jlurgo replied:
Ok I implemented what we talked about, updating mqttSync with running packages status (instead of publishing to mqttClient) and hooking to updates to start/stop logs, status and metrics monitoring

chfritz commented on robot-agent/mqtt.js:
```diff
@@ -180,11 +180,45 @@ mqttClient.on('connect', function(connackPacket) {
+          }
+        }
+      );
+      try {
```
What are these `try-catch`es for, meaning, what could go wrong that needs to be caught?

jlurgo replied:
I was just trying to be extra safe to avoid interrupting mqtt initialization.
The most dangerous of those called methods is logMonitor.watchLogs which does a couple of sync calls to **fs**, processes pending offline logs and starts watching the log file. All this will probably not happen in this call (to watch the agent-itself logs) because it'll be deferred until the next heartbeat.
Should we make watchLogs safer? Always defer log watching initialization for a package?

chfritz commented:
I approved too soon. When I checked out this branch and restarted the agent, mqtt went bananas:

https://github.com/user-attachments/assets/e75f7ffa-ddd8-4ef4-b464-4759f08143af

chfritz reviewed (changes requested):
not sure how to unapprove a PR, so I'll just mark it as request changes; We'll need to debug that mqtt behavior. Looks like an infinite loop of sorts.

chfritz commented:
The issues seems to be the way it uploads existing log files. My agent-test capability prints an `info` line once a second, so the log file is pretty big. These lines seem to be sent in small batches that are not zipped or anything. Zipping them would probably reduce their volume 20x or similar because there is a *ton* of repetition.
```
/cfritz/d_f5b1b62bd4/@transitive-robotics/_robot-agent/2.4.0/status/logs/live [{"timestamp":1755887824135,"module":"main","level":"INFO","logLevelValue":9,"message":"cloud: /cloud/date 2025-08-22T18:37:04.134Z","package":"@local/agent-test"},{"timestamp":1755887825136,"module":"main","level":"INFO","logLevelValue":9,"message":"cloud: /cloud/date 2025-08-22T18:37:05.135Z","package":"@local/agent-test"},{"timestamp":1755887826136,"module":"main","level":"INFO","logLevelValue":9,"message":"cloud: /cloud/date 2025-08-22T18:37:06.135Z","package":"@local/agent-test"},{"timestamp":1755887827137,"module":"main","level":"INFO","logLevelValue":9,"message":"cloud: /cloud/date 2025-08-22T18:37:07.136Z","package":"@local/agent-test"},{"timestamp":1755887828139,"module":"main","level":"INFO","logLevelValue":9,"message":"cloud: /cloud/date 2025-08-22T18:37:08.138Z","package":"@local/agent-test"}] false
/cfritz/d_f5b1b62bd4/@transitive-robotics/_robot-agent/2.4.0/status/logs/lastLogTimestamp 1755887828139 true
/cfritz/d_f5b1b62bd4/@transitive-robotics/_robot-agent/2.4.0/status/logs/live [{"timestamp":1755887829140,"module":"main","level":"INFO","logLevelValue":9,"message":"cloud: /cloud/date 2025-08-22T18:37:09.139Z","package":"@local/agent-test"},{"timestamp":1755887830140,"module":"main","level":"INFO","logLevelValue":9,"message":"cloud: /cloud/date 2025-08-22T18:37:10.139Z","package":"@local/agent-test"},{"timestamp":1755887831141,"module":"main","level":"INFO","logLevelValue":9,"message":"cloud: /cloud/date 2025-08-22T18:37:11.140Z","package":"@local/agent-test"},{"timestamp":1755887832140,"module":"main","level":"INFO","logLevelValue":9,"message":"cloud: /cloud/date 2025-08-22T18:37:12.139Z","package":"@local/agent-test"},{"timestamp":1755887833142,"module":"main","level":"INFO","logLevelValue":9,"message":"cloud: /cloud/date 2025-08-22T18:37:13.141Z","package":"@local/agent-test"}] false
```

chfritz commented:
Given the above finding, I realize that this is unrelated to this PR, so I'll merge and fix the batching logging separately.

chfritz reviewed (approved):
the found issue is unrelated to the changes of this PR, it only made them visible. Will fix them separately.
